File: apd/src/event.rs
use std::{
    env,
    ffi::CStr,
    fs,
    os::unix::{fs::PermissionsExt, process::CommandExt},
    path::{Path, PathBuf},
    process::Command,
    sync::{Arc, Mutex},
    thread,
    time::Duration,
};

use anyhow::{Context, Result};
use libc::SIGPWR;
use log::{info, warn};
use notify::{
    Config, Event, EventKind, INotifyWatcher, RecursiveMode, Watcher,
    event::{ModifyKind, RenameMode},
};
use signal_hook::{consts::signal::*, iterator::Signals};

use crate::{
    assets, defs, metamodule, module, restorecon, supercall,
    supercall::{
        fork_for_result, init_load_package_uid_config, init_load_su_path, refresh_ap_package_list,
    },
    utils::{
        switch_cgroups, {self},
    },
};

pub fn on_post_data_fs(superkey: Option<String>) -> Result<()> {
    utils::umask(0);
    use std::process::Stdio;
    #[cfg(unix)]
    init_load_package_uid_config(&superkey);

    init_load_su_path(&superkey);

    let args = ["/data/adb/ap/bin/magiskpolicy", "--magisk", "--live"];
    fork_for_result("/data/adb/ap/bin/magiskpolicy", &args, &superkey);

    info!("Re-privilege apd profile after injecting sepolicy");
    supercall::privilege_apd_profile(&superkey);

    if utils::has_magisk() {
        warn!("Magisk detected, skip post-fs-data!");
        return Ok(());
    }

    // Create log environment
    if !Path::new(defs::APATCH_LOG_FOLDER).exists() {
        fs::create_dir(defs::APATCH_LOG_FOLDER).expect("Failed to create log folder");
        let permissions = fs::Permissions::from_mode(0o700);
        fs::set_permissions(defs::APATCH_LOG_FOLDER, permissions)
            .expect("Failed to set permissions");
    }
    let command_string = format!(
        "rm -rf {}*.old.log; for file in {}*; do mv \"$file\" \"$file.old.log\"; done",
        defs::APATCH_LOG_FOLDER,
        defs::APATCH_LOG_FOLDER
    );
    let mut args = vec!["-c", &command_string];
    // for all file to .old
    let result = utils::run_command("sh", &args, None)?.wait()?;
    if result.success() {
        info!("Successfully deleted .old files.");
    } else {
        info!("Failed to delete .old files.");
    }
    let logcat_path = format!("{}locat.log", defs::APATCH_LOG_FOLDER);
    let dmesg_path = format!("{}dmesg.log", defs::APATCH_LOG_FOLDER);
    let bootlog = fs::File::create(dmesg_path)?;
    args = vec![
        "-s",
        "9",
        "120s",
        "logcat",
        "-b",
        "main,system,crash",
        "-f",
        &logcat_path,
        "logcatcher-bootlog:S",
        "&",
    ];
    let _ = unsafe {
        Command::new("timeout")
            .process_group(0)
            .pre_exec(|| {
                switch_cgroups();
                Ok(())
            })
            .args(args)
            .spawn()
    };
    args = vec!["-s", "9", "120s", "dmesg", "-w"];
    let _result = unsafe {
        Command::new("timeout")
            .process_group(0)
            .pre_exec(|| {
                switch_cgroups();
                Ok(())
            })
            .args(args)
            .stdout(Stdio::from(bootlog))
            .spawn()
    };

    let key = "KERNELPATCH_VERSION";
    match env::var(key) {
        Ok(value) => println!("{}: {}", key, value),
        Err(_) => println!("{} not found", key),
    }

    let key = "KERNEL_VERSION";
    match env::var(key) {
        Ok(value) => println!("{}: {}", key, value),
        Err(_) => println!("{} not found", key),
    }

    let safe_mode = utils::is_safe_mode(superkey.clone());

    if safe_mode {
        // we should still mount modules.img to `/data/adb/modules` in safe mode
        // becuase we may need to operate the module dir in safe mode
        warn!("safe mode, skip common post-fs-data.d scripts");
        if let Err(e) = module::disable_all_modules() {
            warn!("disable all modules failed: {}", e);
        }
    } else {
        // Then exec common post-fs-data scripts
        if let Err(e) = module::exec_common_scripts("post-fs-data.d", true) {
            warn!("exec common post-fs-data scripts failed: {}", e);
        }
    }
    let module_update_dir = defs::MODULE_UPDATE_DIR; //save module place
    let module_dir = defs::MODULE_DIR; // run modules place
    let module_update_flag = Path::new(defs::WORKING_DIR).join(defs::UPDATE_FILE_NAME); // if update ,there will be renewed modules file
    assets::ensure_binaries().with_context(|| "binary missing")?;

    if Path::new(defs::MODULE_UPDATE_DIR).exists() {
        module::handle_updated_modules()?;
        fs::remove_dir_all(module_update_dir)?;
    }

    if safe_mode {
        warn!("safe mode, skip post-fs-data scripts and disable all modules!");
        if let Err(e) = module::disable_all_modules() {
            warn!("disable all modules failed: {}", e);
        }
        return Ok(());
    }

    if let Err(e) = module::prune_modules() {
        warn!("prune modules failed: {}", e);
    }

    if let Err(e) = restorecon::restorecon() {
        warn!("restorecon failed: {}", e);
    }

    // load sepolicy.rule
    if module::load_sepolicy_rule().is_err() {
        warn!("load sepolicy.rule failed");
    }

    if let Err(e) = metamodule::exec_mount_script(module_dir) {
        warn!("execute metamodule mount failed: {e}");
    }

    // exec modules post-fs-data scripts
    // TODO: Add timeout
    if let Err(e) = module::exec_stage_script("post-fs-data", true) {
        warn!("exec post-fs-data scripts failed: {}", e);
    }
    if let Err(e) = module::exec_stage_lua("post-fs-data", true, superkey.as_deref().unwrap_or(""))
    {
        warn!("Failed to exec post-fs-data lua: {}", e);
    }
    // load system.prop
    if let Err(e) = module::load_system_prop() {
        warn!("load system.prop failed: {}", e);
    }

    info!("remove update flag");
    let _ = fs::remove_file(module_update_flag);

    run_stage("post-mount", superkey, true);

    env::set_current_dir("/").with_context(|| "failed to chdir to /")?;

    Ok(())
}

fn run_stage(stage: &str, superkey: Option<String>, block: bool) {
    utils::umask(0);

    if utils::has_magisk() {
        warn!("Magisk detected, skip {stage}");
        return;
    }

    if utils::is_safe_mode(superkey.clone()) {
        warn!("safe mode, skip {stage} scripts");
        if let Err(e) = module::disable_all_modules() {
            warn!("disable all modules failed: {}", e);
        }
        return;
    }

    if let Err(e) = module::exec_common_scripts(&format!("{stage}.d"), block) {
        warn!("Failed to exec common {stage} scripts: {e}");
    }
    if let Err(e) = module::exec_stage_script(stage, block) {
        warn!("Failed to exec {stage} scripts: {e}");
    }
    if let Err(e) = module::exec_stage_lua(stage, block, superkey.as_deref().unwrap_or("")) {
        warn!("Failed to exec {stage} lua: {e}");
    }

    // execute metamodule stage script first (priority)
    if let Err(e) = metamodule::exec_stage_script(stage, block) {
        warn!("Failed to exec metamodule {stage} script: {e}");
    }

    // execute regular modules stage scripts
    if let Err(e) = crate::module::exec_stage_script(stage, block) {
        warn!("Failed to exec {stage} scripts: {e}");
    }
}

pub fn on_services(superkey: Option<String>) -> Result<()> {
    info!("on_services triggered!");
    run_stage("service", superkey, false);

    Ok(())
}

fn run_uid_monitor() {
    info!("Trigger run_uid_monitor!");

    let mut command = &mut Command::new("/data/adb/apd");
    {
        command = command.process_group(0);
        command = unsafe {
            command.pre_exec(|| {
                // ignore the error?
                switch_cgroups();
                Ok(())
            })
        };
    }
    command = command.arg("uid-listener");

    command
        .spawn()
        .map(|_| ())
        .expect("[run_uid_monitor] Failed to run uid monitor");
}

pub fn on_boot_completed(superkey: Option<String>) -> Result<()> {
    info!("on_boot_completed triggered!");

    run_stage("boot-completed", superkey, false);

    run_uid_monitor();
    Ok(())
}

pub fn start_uid_listener() -> Result<()> {
    info!("start_uid_listener triggered!");
    println!("[start_uid_listener] Registering...");

    // create inotify instance
    const SYS_PACKAGES_LIST_TMP: &str = "/data/system/packages.list.tmp";
    let sys_packages_list_tmp = PathBuf::from(&SYS_PACKAGES_LIST_TMP);
    let dir: PathBuf = sys_packages_list_tmp.parent().unwrap().into();

    let (tx, rx) = std::sync::mpsc::channel();
    let tx_clone = tx.clone();
    let mutex = Arc::new(Mutex::new(()));

    {
        let mutex_clone = mutex.clone();
        thread::spawn(move || {
            let mut signals = Signals::new(&[SIGTERM, SIGINT, SIGPWR]).unwrap();
            for sig in signals.forever() {
                log::warn!("[shutdown] Caught signal {sig}, refreshing package list...");
                let skey = CStr::from_bytes_with_nul(b"su\0")
                    .expect("[shutdown_listener] CStr::from_bytes_with_nul failed");
                refresh_ap_package_list(&skey, &mutex_clone);
                break; // 执行一次后退出线程
            }
        });
    }

    let mut watcher = INotifyWatcher::new(
        move |ev: notify::Result<Event>| match ev {
            Ok(Event {
                kind: EventKind::Modify(ModifyKind::Name(RenameMode::Both)),
                paths,
                ..
            }) => {
                if paths.contains(&sys_packages_list_tmp) {
                    info!("[uid_monitor] System packages list changed, sending to tx...");
                    tx_clone.send(false).unwrap()
                }
            }
            Err(err) => warn!("inotify error: {err}"),
            _ => (),
        },
        Config::default(),
    )?;

    watcher.watch(dir.as_ref(), RecursiveMode::NonRecursive)?;

    let mut debounce = false;
    while let Ok(delayed) = rx.recv() {
        if delayed {
            debounce = false;
            let skey = CStr::from_bytes_with_nul(b"su\0")
                .expect("[start_uid_listener] CStr::from_bytes_with_nul failed");
            refresh_ap_package_list(&skey, &mutex);
        } else if !debounce {
            thread::sleep(Duration::from_secs(1));
            debounce = true;
            tx.send(true)?;
        }
    }

    Ok(())
}

File: apd_go/event.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
)

func execCommand(command string, args []string) error {
	cmd := exec.Command(command, args...)
	return cmd.Run()
}
func on_post_fs_data(superkey string) {
	Umask(0)

	InitLoadPackageUidConfig(superkey)
	InitLoadSuPath(superkey)

	args := []string{"--magisk", "--live"}
	if err := execCommand(magiskpolicy, args); err != nil {
		Error("Load magiskpolicy failed")
	}

	Info("Re-privilege apd profile after injecting sepolicy")
	PrivilegeApdProfile(superkey)

	if HasMagisk() {
		Error("Magisk detected, skip post-fs-data!")
		return
	}

	// Create log environment
	if _, err := os.Stat(ap_log); os.IsNotExist(err) {
		if err := os.Mkdir(ap_log, 0700); err != nil {
			Error("failed to create log folder: %w", err)
		}
	}

	// Remove old log files
	commandString := fmt.Sprintf("rm -rf %s*.old; for file in %s*; do mv \"$file\" \"$file.old\"; done", ap_log, ap_log)

	if err := execCommand("sh", []string{"-c", commandString}); err != nil {
		Warn("failed to delect old log")
	}

	logcatPath := filepath.Join(ap_log, "logcat.log")
	dmesgLog := filepath.Join(ap_log, "dmesg.log")
	cmdStr := fmt.Sprintf("nohup timeout -s 9 120s logcat -b main,system,crash -f %s logcatcher-bootlog:S &", logcatPath)
	cmd := exec.Command("/bin/sh", "-c", cmdStr)
	if err := cmd.Run(); err != nil {
		Error("failed to start logcat: %v\n", err)
	}

	cmdStr = fmt.Sprintf("nohup timeout -s 9 120s dmesg -w > %s 2>&1 &", dmesgLog)
	cmd = exec.Command("/bin/sh", "-c", cmdStr)
	if err := cmd.Run(); err != nil {
		Error("failed to start dmesg: %v\n", err)
	}
	safeMode := isSafeMode(&superkey)

	if safeMode {
		if err := disableAllModulesUpdate(); err != nil {
			Error("disable all modules failed: %v", err)
		}
	} else {
		if err := execCommonScripts("post-fs-data.d", true); err != nil {
			//warn(fmt.Sprintf("exec common post-fs-data scripts failed: %v", err))
			Error("exec common post-fs-data scripts failed: %v", err)
		}
	}

	moduleUpdateFlag := filepath.Join(workingDir, updateFileName)
	if err := ensureBinary(binaryDir); err != nil {
		Error("binary missing: %w", err)
		return
	}
	if _, err := os.Stat(moduleupdateDir); err == nil {
		if err := moveFile(moduleupdateDir, moduleDir); err != nil {

		}
		if err := os.RemoveAll(moduleupdateDir); err != nil {

		}
	}
	tmpModuleImg := tmp_img
	tmpModulePath := filepath.Join(tmpModuleImg)

	if (fileExists(moduleUpdateFlag) || !fileExists(tmpModulePath)) && shouldEnableOverlay() {
		//todo

	} else {
		//info("donothing")
	}

	if safeMode {
		//warn("safe mode, skip post-fs-data scripts and disable all modules!")
		if err := disableAllModulesUpdate(); err != nil {
			//warn(fmt.Sprintf("disable all modules failed: %v", err))
			Error("disable all modules failed: %v", err)
		}
		return
	}

	if err := pruneModules(); err != nil {
		Error("prune modules failed: %v", err)
	}

	if err := RestoreCon(); err != nil {
		Error("restorecon failed: %v", err)
	}

	if err := loadSEPolicyRule(); err != nil {
		Error("load sepolicy.rule failed")
	}

	if err := mountTmpfs(getTmpPath()); err != nil {
		Error("do temp dir mount failed: %v", err)
	}

	// Execute modules post-fs-data scripts
	if err := ExecStageScript("post-fs-data", true); err != nil {
		Error("exec post-fs-data scripts failed: %v", err)
	}

	// Load system.prop
	if err := loadSystemProp(); err != nil {
		Error("load system.prop failed: %v", err)
	}
	//magicMount()
	//if shouldEnableOverlay() {
	//	if err := mountSystemlessly(moduleDir); err != nil {
	//		warn(fmt.Sprintf("do systemless mount failed: %v", err))
	//	}
	//} else {
	//	if err := systemlessBindMount(moduleDir); err != nil {
	//		warn(fmt.Sprintf("do systemless bind_mount failed: %v", err))
	//	}
	//}

	runStage("post-mount", &superkey, true)

	if err := os.Chdir("/"); err != nil {
		Error("failed to chdir to /: %w", err)
	}

	return
}
func on_services(superkey string) {
	Info("on_services triggered!")
	runStage("service", &superkey, false)
}
func on_boot_completed(superkey string) {
	Info("on_boot_completed triggered!")
	runStage("boot-completed", &superkey, false)
}
func runStage(stage string, superkey *string, block bool) {
	Umask(0)

	if HasMagisk() {
		Info("Magisk detected, skip %s", stage)
		return
	}

	if isSafeMode(superkey) {
		Info("safe mode, skip %s scripts", stage)
		if err := disableAllModulesUpdate(); err != nil {
			Error("disable all modules failed: %v", err)
		}
		return
	}

	if err := execScript(fmt.Sprintf("%s.d", stage), block); err != nil {
		Error("Failed to exec common %s scripts: %v", stage, err)
	}
	if err := ExecStageScript(stage, block); err != nil {
		Error("Failed to exec %s scripts: %v", stage, err)
	}
}

func moveFile(moduleUpdateDir, moduleDir string) error {
	entries, err := os.ReadDir(moduleUpdateDir)
	if err != nil {
		return err
	}

	for _, entry := range entries {
		fileName := entry.Name()
		sourcePath := filepath.Join(moduleUpdateDir, fileName)
		targetPath := filepath.Join(moduleDir, fileName)

		if entry.IsDir() {
			updateFilePath := filepath.Join(targetPath, updateFileName)
			if _, err := os.Stat(updateFilePath); err == nil {
				if _, err := os.Stat(targetPath); err == nil {
					Info("Removing existing folder in target directory: %s\n", fileName)
					if err := os.RemoveAll(targetPath); err != nil {
						return err
					}
				}

				Info("Moving %s to target directory\n", fileName)
				if err := os.Rename(sourcePath, targetPath); err != nil {
					return err
				}
			}
		}
	}

	return nil
}
